---
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%"
)

set.seed(4394)
```

# Rank-gap statistics for exploring high-dimensional overlap

<!-- badges: start -->
<!-- badges: end -->

This package contains functions for computing rank-gap and related statistics,
and plotting the results. These statistics provide a convenient way to
represent greater-than-expected concordance among items ranked highly across
all of number of lists.

Rank-gap statistics were developed to understand "overlap" and in genomics and
high throughput biology, and in particular to complement the widespread use of
Venn diagrams for this purpose when errors are independent across experiments
or conditions. But we expect they may be useful for exploring multivariate
distributions in other contexts.


## Example

```{r}
library(rankGap)
```

We illustrate the package using data from a comparison of (disease vs.
wild-type) gene expression differences in mice with three relate genetic
diseases (Mendelian Disorders of the Epigenetic Machinery, or MDEMs).

Importantly, diseased mice were compared with littermate controls, so we expect
technical errors across the three (disease vs. wilde-type) comparisons to be
independent.

The data are from (Luperchio et. al., 2021); `d_B_limma` contains output from a
limma reanalysis of the raw read counts:

```{r}
tibble::tibble(d_B_limma) |> head(3)
```

Let's begin by a visualization of the distribution of rank-gap statistics across
genes:

```{r fig.height=3, fig.width=4, out.width="45%"}
with(d_B_limma, rank_gap_hist(t.KS1, t.KS2, t.RT))
```

The `rank_gap_hist` function takes as input _signed_ statistics—here
_t_-values— and generates within-condition ranks and the associated rank-gap
statistics from their absolute values.

Rank-gap statistics have *p*-value-like properties: they are (approximately)
uniformly distributed when gene ranks are independent, and smaller-than-uniform
when gene ranks are closer than expected by chance.

Thus, the spike near 0 in the distribution indicates **greater-than-expected
concordance in the ranks of (absolute) *t*-scores for a given gene across
conditions**.

The histogram has been stratified by the signs of the _t_-scores, revealing
that the excess concordance is driven by genes with the same signs in all 3
comparisons—i.e., either upregulated in all 3 MDEMs or downregulated in all 3
MDEMs.

As a measure of concordance, rank-gap statistics implicitly emphasize genes
that are ranked high in any of the 3 conditions—that is, for which the maximum
of the ranks is large. However we can more explicitly focus on high-ranked
genes using the fact that rank-gap statistics are also uniformly distributed
conditional on the maximum rank:

```{r fig.height=3, fig.width=6.5, out.width="60%"}
with(d_B_limma, rank_gap_stephist(t.KS1, t.KS2, t.RT, n_max_rank_bins = 4))
```

Here you can see that the signal of overlap is concentrated among the top 25%
of genes (ordered by maximum rank across the 3 conditions or, loosely,
"importance in any condition").

(We use "line histograms" here to avoid overplotting.)

### Comparison with Venn diagrams of significant genes

Let's compare this with the standard approach in computational biology of
computing Venn diagrams of (margianlly) statistically significant genes (with
_q_-values < 0.2 as estimated by `qvalue`):

```{r fig.width=4, fig.height=3, out.width="35%"}
require(eulerr)
require(qvalue)
require(ggplot2)

# Matrix of detected DE indicators
m_DE <-
  d_B_limma[c("P.Value.KS1", "P.Value.KS2", "P.Value.RT")] |>
  as.matrix() |>
  apply(2, \(p) qvalue::qvalue(p)$qvalues) |>
  (\(x) x < .2)()

colnames(m_DE) <- c("KS1", "KS2", "RT")

venn(m_DE) |>
  plot(quantities = TRUE, fills = FALSE) |>
  cowplot::plot_grid() +
  ggplot2::theme(plot.margin = ggplot2::margin(10, 10, 10, 10))
```

There are very few genes detected as differentially expressed in all 3 
conditions.

## Installation

You can install the development version of `rankGap` from
[GitHub](https://github.com/) with:

``` r
# install.packages("pak")
pak::pak("l-d-s/rankGap")
```

### Citation

The method was developed in Leon Di Stefano's PhD thesis:

```
@thesis{stefanoTamingInteractionsGenomics2024,
  title = {Taming Interactions in Genomics and Clinical Trials},
  author = {Di Stefano, Leon S},
  date = {2024-09-04},
  institution = {Johns Hopkins University},
  url = {https://jscholarship.library.jhu.edu/handle/1774.2/70173},
  urldate = {2025-01-08},
  abstract = {This dissertation develops statistical methods to address two challenges in contemporary biomedical research. The first is understanding shared mechanisms underlying related diseases using high- throughput molecular data, motivated by a study of gene expression changes in a group of rare genetic disorders called Mendelian Disorders of the Epigenetic Machinery or MDEMs (Luperchio et al., 2021). We develop a simple approach to assessing the degree of overlap among the disorders based on rescaled differences in ranks of each gene across conditions, which we call “rank-gap” statistics. Rank-gap statistics have p-value-like properties that highlight concordance among genes ranked highly in at least one condition, and we argue that the number of smaller than expected rank-gap statistics estimates a meaningful overlap metric in the limit of small measurement error. Rank-gap statistics also possess three- and higher-way analogues. We compare our proposal with other rank-based overlap methods in the literature, and evaluate a diverse group of related methods in a simulation study. Our approach provides a simple complement to the predominant practice in genomics of forming Venn diagrams of statistically significant genes. The second challenge is understanding how treatment effects vary among different types of patients using data from clinical trials. The standard approach—one-at-a-time subgroup or interaction analysis of candidate covariates—suffers from difficulties of inference and interpretation. We develop the proportional interaction model, a parsimonious extension of additive regression adjustment also considered by Follmann and Proschan (1999) and Kovalchik, Varadhan, and Weiss (2013) in which benefits and harms from treatment vary as a function of a “score” that also predicts outcomes in each treatment group. An asymptotic analysis enables us to assess the sample sizes required to detect proportional interactions in practice, and we propose a symmetrical re-parametrization of the model that allows for stable single-step inference. We re-analyze a clinical trial of treatments for schizophrenia using our approach, showing how proportional interactions can be incorporated into a traditional regression model-building framework.},
  langid = {american}
}
```

### References

1. Luperchio TR, Boukas L, Zhang L, Pilarowski G, Jiang J, Kalinousky A, et al.
*Leveraging the Mendelian disorders of the epigenetic machinery to
systematically map functional epigenetic variation*. Dekker J, Barkai N,
editors. eLife. 2021 Aug 31;10:e65884. doi: 10.7554/eLife.65884
